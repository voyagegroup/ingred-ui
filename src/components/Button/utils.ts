import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
