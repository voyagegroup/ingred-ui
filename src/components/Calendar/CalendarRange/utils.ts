import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
