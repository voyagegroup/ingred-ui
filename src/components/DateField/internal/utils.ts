export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
