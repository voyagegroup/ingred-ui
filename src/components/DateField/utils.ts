import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
