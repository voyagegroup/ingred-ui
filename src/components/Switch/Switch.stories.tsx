import * as React from "react";
import { Story } from "@storybook/react/types-6-0";
import Switch from "./Switch";

export default {
  title: "Components/Inputs/Switch",
  component: Switch,
};

export const Example = () => {
  const [index, setIndex] = React.useState(0);
  return (
    <Switch
      value={index}
      cases={[
        { name: "One" },
        { name: "Two" },
        { name: "Three" },
        { name: "Four" },
      ]}
      onChange={setIndex}
    />
  );
};

export const WithIcon = () => {
  const [index, setIndex] = React.useState(0);
  return (
    <Switch
      value={index}
      cases={[
        { name: "Alert", icon: "alart" },
        { name: "Auction", icon: "auction" },
        { name: "Dashboard", icon: "dashboard" },
      ]}
      onChange={setIndex}
    />
  );
};

export const WithValue: Story = () => {
  const [value, setValue] = React.useState<number>(1);
  return (
    <Switch
      value={value}
      cases={[
        { name: "One", value: 1 },
        { name: "Two", value: 2 },
        { name: "Three", value: 3 },
        { name: "Four", value: 4 },
      ]}
      onChange={setValue}
    />
  );
};

WithValue.parameters = {
  docs: {
    description: {
      story: "`cases` props can include `value`.",
    },
  },
};
