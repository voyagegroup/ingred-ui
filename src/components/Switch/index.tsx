export { default, SwitchProps } from "./Switch";
