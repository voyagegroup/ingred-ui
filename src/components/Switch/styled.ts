import styled from "styled-components";

export const Container = styled.div`
  display: inline-flex;
  height: 40px;
  padding: ${({ theme }) => theme.spacing / 2}px;
  border-radius: ${({ theme }) => theme.radius}px;
  background-color: ${({ theme }) => theme.palette.gray.light};
`;

export const ChildContainer = styled.div`
  position: relative;
`;

export const ChildrenContainer = styled.div`
  position: relative;
  width: 100%;
  white-space: nowrap;
  display: flex;
  align-items: center;
`;

export const IconItemContainer = styled.div`
  padding: ${({ theme }) => theme.spacing / 2}px;
  height: 32px;
  cursor: pointer;
  transition: all 0.3s ease-in-out;
  & + & {
    margin-left: ${({ theme }) => theme.spacing / 2}px;
  }
`;

export const TextItemContainer = styled.div`
  display: inline-flex;
  align-items: center;
  height: 32px;
  padding: 0 ${({ theme }) => theme.spacing}px;
  border-radius: ${({ theme }) => theme.radius}px;
  cursor: pointer;
  transition: all 0.3s ease-in-out;
  & + & {
    margin-left: ${({ theme }) => theme.spacing / 2}px;
  }
`;

export const Indicator = styled.div`
  position: absolute;
  top: 0;
  height: 100%;
  transition: all 0.3s ease-in-out;
  background-color: ${({ theme }) => theme.palette.background.default};
  border-radius: ${({ theme }) => theme.radius * 0.5}px;
`;
