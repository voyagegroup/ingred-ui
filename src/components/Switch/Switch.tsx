import * as React from "react";
import * as Styled from "./styled";
import Icon from "../Icon";
import { IconName } from "../Icon/Icon";
import Typography from "../Typography";
import Tooltip from "../Tooltip";
import useEventCallback from "../../hooks/useEventCallback";

type Case = {
  name: string;
  value?: any;
};
type CaseWithIcon = Case & {
  icon: IconName;
};

export type SwitchProps = {
  cases: Case[] | CaseWithIcon[];
  value: any;
  onChange?: (value: any) => void;
};

export type IndicatorStyle = {
  width?: number;
  left?: number;
};

const Switch = React.forwardRef<HTMLDivElement, SwitchProps>(function Switch(
  { cases, value, onChange },
  ref,
) {
  const items = cases as {
    name: string;
    icon?: IconName;
    value?: any;
  }[];
  const valueToIndex = new Map();
  const tabsRef = React.useRef<HTMLDivElement>(null);
  const childrenWrapperRef = React.useRef<HTMLDivElement>(null);
  const [indicatorStyle, setIndicatorStyle] = React.useState<IndicatorStyle>(
    {},
  );
  const handleChange = (value: any) => () => {
    if (onChange) {
      onChange(value);
    }
  };

  const getTabsMeta = () => {
    const tabsNode = tabsRef.current;
    let tabsMeta;
    if (tabsNode) {
      const rect = tabsNode.getBoundingClientRect();
      tabsMeta = {
        clientWidth: tabsNode.clientWidth,
        scrollLeft: tabsNode.scrollLeft,
        scrollTop: tabsNode.scrollTop,
        scrollWidth: tabsNode.scrollWidth,
        top: rect.top,
        bottom: rect.bottom,
        left: rect.left,
        right: rect.right,
      };
    }

    let tabMeta;
    if (tabsNode && value !== false) {
      if (childrenWrapperRef.current) {
        const children = childrenWrapperRef.current.children;

        if (children.length > 0) {
          const tab = children[valueToIndex.get(value)];
          tabMeta = tab?.getBoundingClientRect() || null;
        }
      }
    }
    return { tabsMeta, tabMeta };
  };

  const updateIndicatorState = useEventCallback(() => {
    const { tabsMeta, tabMeta } = getTabsMeta();
    let startValue = 0;

    if (tabMeta && tabsMeta) {
      const correction = tabsMeta.scrollLeft;
      startValue = tabMeta.left - tabsMeta.left + correction;
    }

    const newIndicatorStyle = {
      left: startValue,
      width: tabMeta ? tabMeta.width : 0,
    };

    if (
      indicatorStyle.left === undefined ||
      indicatorStyle.width === undefined
    ) {
      setIndicatorStyle(newIndicatorStyle);
    } else {
      const dStart = Math.abs(indicatorStyle.left - newIndicatorStyle.left);
      const dSize = Math.abs(indicatorStyle?.width - newIndicatorStyle.width);

      if (dStart >= 1 || dSize >= 1) {
        setIndicatorStyle(newIndicatorStyle);
      }
    }
  });

  React.useEffect(() => {
    updateIndicatorState();
  });

  React.useEffect(() => {
    const handleResize = () => {
      updateIndicatorState();
    };

    window.addEventListener("resize", handleResize);
    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, [updateIndicatorState]);

  return (
    <Styled.Container ref={ref}>
      <Styled.ChildContainer ref={tabsRef}>
        <Styled.Indicator
          style={{
            ...indicatorStyle,
          }}
        />
        <Styled.ChildrenContainer ref={childrenWrapperRef}>
          {items.map((item, index) => {
            const childValue = item.value === undefined ? index : item.value;
            valueToIndex.set(childValue, index);
            const active = childValue === value;
            return (
              <Tooltip
                key={item.name}
                content={item.name}
                positionPriority={["top-start", "bottom-start", "auto"]}
                offset={[0, 15]}
              >
                {item.icon != null ? (
                  <Styled.IconItemContainer onClick={handleChange(childValue)}>
                    <Icon name={item.icon} size="lg" />
                  </Styled.IconItemContainer>
                ) : (
                  <Styled.TextItemContainer
                    key={item.name}
                    onClick={handleChange(childValue)}
                  >
                    <Typography
                      component="span"
                      size="sm"
                      weight={active ? "bold" : "normal"}
                      color="secondary"
                    >
                      {item.name}
                    </Typography>
                  </Styled.TextItemContainer>
                )}
              </Tooltip>
            );
          })}
        </Styled.ChildrenContainer>
      </Styled.ChildContainer>
    </Styled.Container>
  );
});

export default Switch;
