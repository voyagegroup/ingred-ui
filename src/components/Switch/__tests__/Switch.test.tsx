import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import Switch from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Switch component testing", () => {
  afterEach(cleanup);

  test("Switch", () => {
    const { asFragment } = renderWithThemeProvider(
      <Switch
        cases={[
          {
            name: "Hoge",
          },
          {
            name: "Fuga",
          },
        ]}
        value={0}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
