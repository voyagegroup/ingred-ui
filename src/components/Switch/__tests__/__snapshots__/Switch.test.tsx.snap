// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch component testing Switch 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL bvyKru"
  >
    <div
      class="sc-gsDKAQ dFzOKQ"
    >
      <div
        class="sc-jRQBWg bhlQaL"
        style="left: 0px; width: 0px;"
      />
      <div
        class="sc-dkPtRN gAIaqV"
      >
        <div
          class="sc-eCImPb VoreP"
        >
          <span
            class="sc-iCfMLu eAWLJn"
            color="#596978"
            font-size="13px"
          >
            Hoge
          </span>
        </div>
        <div
          class="sc-eCImPb VoreP"
        >
          <span
            class="sc-iCfMLu droGrB"
            color="#596978"
            font-size="13px"
          >
            Fuga
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
